### PR TITLE
Handling https url rewrite (proxy mode) + adding getter and setter for some classes

### PR DESCRIPTION
**Pull request by scheylord**
Hello,
We have some internal classes which are extending AccessPoint.java, DefaultReplayCaptureSelector.java and TextReplayRenderer.java.
In order to migrate to openwayback 2.2 (we are currently using version 2.0) we need more visibity on some variables, so if it's possible we would like to add some getters and setters.

### CODE REVIEW

**Comment by scheylord** [comment]
This is an implementation to correct this issue :
https://github.com/iipc/openwayback/issues/25

Basically, we are adding two new classes :

**HttpsRedirectRewriteReplayRendeder** extending TransparentReplayRenderer :

Before giving back the response to the client, we check if the http response header is a redirection (by checking if parameter "Location" or "location" is present) and if the redirect url start with https, we rewrite it.
Particular case : if the redirect link is exactly the same as the original link + https (eg http://www.example.com/ redirect to https://www.example.com/) it would cause a redirect loop, so we handle this case in particular, moving to the next of previous capture.

**HttpsRedirectAndLinksRewriteProxyHTMLMarkupReplayRenderer** extending ProxyHTMLMarkupReplayRenderer :

Do the same thing as HttpsRedirectRewriteReplayRendeder, but then we also process to transform all the links we could find in the page content.
(if the mime type is text/html as define in ProxyReplay.xml)

**ProxyReplay.xml**

Replace the classes with their extended one (HttpsRedirectAndLinksRewriteProxyHTMLMarkupReplayRenderer replaces ProxyHTMLMarkupReplayRendere and HttpsRedirectRewriteReplayRendeder  replaces TransparentReplayRenderer ).

**wayback.xml (not in the commit)**

Finally in your wayback.xml configuration file, you also have to add to your proxy replay access point :


<property name="exactSchemeMatch" value="false"/>

File: wayback-webapp/src/main/webapp/WEB-INF/ProxyReplay.xml [hunk 1]
```diff
@@ -6,9 +6,13 @@
 
   <bean id="identityhttpheaderprocessor" class="org.archive.wayback.replay.IdentityHttpHeaderProcessor" />
 
-  <bean id="proxytransparentreplayrenderer" class="org.archive.wayback.replay.TransparentReplayRenderer">
+  <!-- use org.archive.wayback.replay.HttpsRedirectRewriteReplayRendeder instead of
+  org.archive.wayback.replay.TransparentReplayRendeder to force https url rewrite -->
+  <bean id="proxytransparentreplayrenderer" class="org.archive.wayback.replay.TransparentReplayRendeder">
```
Comment (ldko) [inline wayback-webapp/src/main/webapp/WEB-INF/ProxyReplay.xml]: If this remains the default,
 class="org.archive.wayback.replay.TransparentReplayRendeder"
should be
 class="org.archive.wayback.replay.TransparentReplayRende**rer**"

Comment (scheylord) [inline wayback-webapp/src/main/webapp/WEB-INF/ProxyReplay.xml]: I've committed the correction, thank you Lauren.

File: wayback-core/src/main/java/org/archive/wayback/replay/HttpsRedirectRewriteReplayRendeder.java [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+import org.archive.wayback.core.CaptureSearchResults;
+import org.archive.wayback.core.Resource;
+import org.archive.wayback.core.WaybackRequest;
+import org.archive.wayback.exception.BadContentException;
+
+/**
+ * rewrite https link to http in http header (redirect)
+ * @author lam
+ */
+public class HttpsRedirectRewriteReplayRendeder extends
```
Comment (anjackson) [inline wayback-core/src/main/java/org/archive/wayback/replay/HttpsRedirectRewriteReplayRendeder.java]: Typo here: HttpsRedirectRewriteReplayRendeder should be HttpsRedirectRewriteReplayRenderer

**Comment by anjackson** [comment]
This is potentially a bit of overkill, as there are existing hooks to handle most of this. If you declare an appropriate uriConverter on the AccessPoint, then the HTTPS > HTTP rewriting of location headers is already handled. 

See https://github.com/ukwa/docker-openwayback/blob/master/wayback.xml#L344-L346

The case of rewriting HTTPS links inside the document is not currently handled properly by org.archive.wayback.proxy.ProxyHTMLMarkupReplayRenderer, but I think the same could be achieved by extending that class rather than reproducing all the logic of the parent class.

**Comment by ldko** [comment]
Closing this, as we've decided not to take this approach for https link rewriting.

**Comment by csrster** [comment]
What is the status of https link-rewriting now? The lack of a fix for this issue is/was blocking for adoption of openwayback at the Danish netarkivet.
--Colin

**Comment by ldko** [comment]
The example of a `uriConverter` on the `AccessPoint` to handle the HTTPS > HTTP rewriting of location headers link in the @anjackson  [comment](https://github.com/iipc/openwayback/pull/294#issuecomment-194813510) above can now be seen via 
https://github.com/ukwa/docker-openwayback/blob/14b43463e47a45ce99de66d3cee56f76856e30c7/wayback.xml#L344-L346

@csrster Regarding rewriting HTTPS links inside the document, work has not been done to fix that. Is it something your group could contribute work on?

@anjackson in order to get someone started on implementing a fix for this, would you be willing to expand on how `org.archive.wayback.proxy.ProxyHTMLMarkupReplayRenderer` could be extended to handle rewriting HTTPS links?

**Pull request closed without merging**